File: ui/src/types/events.ts
import { z } from 'zod'

/**
 * Execution event contracts — what the backend streams over WebSocket.
 *
 * Mirrors agentforge.md's core event structure. `runId`, `callIndex`,
 * `parentAgent`, and `policyDecision` are present from the start so that
 * v0.3 (local re-run, agent→agent flow) and v0.5 (branching) layer on without
 * changing the model. See ui-architecture.md §4.3.
 */

export const EXECUTION_EVENT_TYPES = [
  'node_start',
  'node_end',
  'tool_call',
  'agent_action',
  'log',
  'error',
  // backend-emitted
  'policy_decision',
  'token_usage',
  'decision_record',
  'interrupt',
] as const

export const TokenUsageSchema = z.object({
  prompt: z.number(),
  completion: z.number(),
  /** 실제 사용된 모델명 (관측용) */
  model: z.string().optional(),
  /** 성공한 시도의 인덱스 (0=첫 시도) — 재시도 횟수 표시 */
  attempt: z.number().optional(),
  /** fallback 모델로 처리됐는지 */
  fallbackUsed: z.boolean().optional(),
})
export type TokenUsage = z.infer<typeof TokenUsageSchema>

/** Why a review/policy node activated or skipped — v0.1's key portfolio signal. */
export const PolicyDecisionSchema = z.object({
  activated: z.boolean(),
  reason: z.string(),
  /** review 노드 판정: accept / refine / clarify */
  branch: z.string().optional(),
  /** batch_mode에서 clarify → accept 강등 여부 */
  demoted: z.boolean().optional(),
})
export type PolicyDecision = z.infer<typeof PolicyDecisionSchema>

/** Structured failure detail — backend make_error_event() {type, detail}. */
export const ExecutionErrorSchema = z.object({
  type: z.string(),
  detail: z.string(),
})
export type ExecutionError = z.infer<typeof ExecutionErrorSchema>

// Only 'maxIterations' | 'budget' | 'stuck' are reachable on the wire today —
// server/nodes/loop_guard.py's evaluate_loop_guard never emits 'success' (it
// omits exitReason entirely when should_continue is true) or 'escalated', and
// the 'fail' onExhaustion path raises before any event can carry 'failed'.
// Keeping the unreachable values in the enum is a deliberate reserved-for-later
// choice, not an oversight — see server/graphs/compile.py's loop.guard branch.
export const LOOP_RUNTIME_EXIT_REASONS = [
  'success',
  'maxIterations',
  'budget',
  'stuck',
  'escalated',
  'failed',
] as const

/** Per-pass snapshot emitted by a `loop.guard` canvas node — see
 * docs/superpowers/specs/2026-08-07-loop-node-design.md §8. `loopNodeId` is the
 * real canvas node id, identical to the event's own `nodeId`, so consumers can
 * treat guard events exactly like any other node's events. */
export const LoopRuntimeSchema = z.object({
  loopNodeId: z.string(),
  iteration: z.number(),
  maxIterations: z.number().optional(),
  tokens: z.number().optional(),
  costUsd: z.number().optional(),
  durationMs: z.number().optional(),
  lastFeedback: z.string().optional(),
  exitReason: z.enum(LOOP_RUNTIME_EXIT_REASONS).optional(),
})
export type LoopRuntime = z.infer<typeof LoopRuntimeSchema>

export const ExecutionEventSchema = z.object({
  eventType: z.enum(EXECUTION_EVENT_TYPES),
  runId: z.string(),
  nodeId: z.string(),
  agentId: z.string().optional(),
  /** Who triggered this node — drives agent→agent flow view (v0.3). */
  parentAgent: z.string().optional(),
  /** Nth invocation of this node within the run — edge thickness (v0.3). */
  callIndex: z.number().optional(),
  durationMs: z.number().optional(),
  input: z.unknown().optional(),
  output: z.unknown().optional(),
  tokenUsage: TokenUsageSchema.optional(),
  policyDecision: PolicyDecisionSchema.optional(),
  loopRuntime: LoopRuntimeSchema.optional(),
  message: z.string().optional(),
  error: ExecutionErrorSchema.optional(),
  timestamp: z.string(),
})
export type ExecutionEvent = z.infer<typeof ExecutionEventSchema>

/** Runtime status derived from events by the eventReducer (not sent raw). */
export type NodeRuntimeStatus =
  | 'idle'
  | 'running'
  | 'success'
  | 'failed'
  | 'skipped'

export type NodeRuntime = {
  status: NodeRuntimeStatus
  callCount: number
  totalDurationMs: number
  totalTokens: number
  lastInput?: unknown
  lastOutput?: unknown
  policyDecision?: PolicyDecision
  lastError?: ExecutionError
  /** 마지막 성공 호출의 모델명 (관측용) */
  lastModel?: string
  /** 마지막 성공 호출의 시도 인덱스 (0=첫 시도) — 재시도 표시 */
  lastAttempt?: number
  /** 마지막 호출이 fallback 모델로 처리됐는지 */
  fallbackUsed?: boolean
}

export const EMPTY_RUNTIME: NodeRuntime = {
  status: 'idle',
  callCount: 0,
  totalDurationMs: 0,
  totalTokens: 0,
}

File: server/events.py
from datetime import UTC, datetime
from typing import Any, Protocol

# ─── Event model (frontend-compatible flat shape) ──────────────────────────────


class ExecutionEvent:
    """
    프론트 ExecutionEventSchema와 1:1 대응.
    serialize()가 WebSocket으로 나가는 camelCase dict를 반환한다.
    """

    __slots__ = (
        "event_type",
        "run_id",
        "node_id",
        "timestamp",
        "duration_ms",
        "input",
        "output",
        "token_usage",
        "policy_decision",
        "loop_runtime",
        "message",
        "error",
    )

    def __init__(
        self,
        event_type: str,
        run_id: str,
        node_id: str,
        *,
        duration_ms: int | None = None,
        input: Any = None,
        output: Any = None,
        token_usage: dict | None = None,
        policy_decision: dict | None = None,
        loop_runtime: dict | None = None,
        message: str | None = None,
        error: dict | None = None,
    ) -> None:
        self.event_type = event_type
        self.run_id = run_id
        self.node_id = node_id
        self.timestamp = datetime.now(UTC).isoformat()
        self.duration_ms = duration_ms
        self.input = input
        self.output = output
        self.token_usage = token_usage
        self.policy_decision = policy_decision
        self.loop_runtime = loop_runtime
        self.message = message
        self.error = error

    def to_frontend(self) -> dict:
        """프론트 ExecutionEventSchema와 일치하는 camelCase dict."""
        d: dict[str, Any] = {
            "eventType": self.event_type,
            "runId": self.run_id,
            "nodeId": self.node_id,
            "timestamp": self.timestamp,
        }
        if self.duration_ms is not None:
            d["durationMs"] = self.duration_ms
        if self.input is not None:
            d["input"] = self.input
        if self.output is not None:
            d["output"] = self.output
        if self.token_usage is not None:
            d["tokenUsage"] = self.token_usage
        if self.policy_decision is not None:
            d["policyDecision"] = self.policy_decision
        if self.loop_runtime is not None:
            d["loopRuntime"] = self.loop_runtime
        if self.message is not None:
            d["message"] = self.message
        if self.error is not None:
            d["error"] = self.error
        return d

    def model_dump(self) -> dict:
        """harness용 snake_case dict (내부 직렬화). to_frontend()와 필드 대칭 유지."""
        return {
            "event_type": self.event_type,
            "run_id": self.run_id,
            "node_id": self.node_id,
            "timestamp": self.timestamp,
            "duration_ms": self.duration_ms,
            "input": self.input,
            "output": self.output,
            "token_usage": self.token_usage,
            "policy_decision": self.policy_decision,
            "loop_runtime": self.loop_runtime,
            "message": self.message,
            "error": self.error,
        }


class EventEmitter(Protocol):
    async def __call__(self, event: ExecutionEvent) -> None: ...


def make_event(
    run_id: str,
    node_id: str,
    event_type: str,
    **kwargs: Any,
) -> ExecutionEvent:
    return ExecutionEvent(event_type=event_type, run_id=run_id, node_id=node_id, **kwargs)


def make_error_event(
    run_id: str,
    node_id: str,
    exc: BaseException,
    *,
    message: str | None = None,
) -> ExecutionEvent:
    """노드 실패 표준 이벤트. event_type="error" + 구조화된 error 필드.

    프론트 eventReducer가 'error' → status 'failed'로 매핑한다.
    error 필드는 {type, detail} 형태로 LogPanel/Inspector에 그대로 표시된다.
    """
    return ExecutionEvent(
        event_type="error",
        run_id=run_id,
        node_id=node_id,
        message=message or str(exc),
        error={"type": type(exc).__name__, "detail": str(exc)},
    )


# ─── WebSocket emitter ─────────────────────────────────────────────────────────


class WSEventEmitter:
    def __init__(self, websocket: Any, run_id: str, history: list[ExecutionEvent]):
        self.ws = websocket
        self.run_id = run_id
        self.history = history

    async def __call__(self, event: ExecutionEvent) -> None:
        self.history.append(event)
        # node_end 이벤트의 output을 워크스페이스에 .md 파일로 영속화
        if event.event_type == "node_end" and event.output is not None:
            from workspace import write_node_output

            write_node_output(self.run_id, event.node_id, event.output)
        await self.ws.send_json({"kind": "event", "event": event.to_frontend()})


# ─── List emitter (for harness) ────────────────────────────────────────────────


class ListEventEmitter:
    def __init__(self) -> None:
        self.events: list[ExecutionEvent] = []

    async def __call__(self, event: ExecutionEvent) -> None:
        self.events.append(event)
